From 59e9f5a34cb3df3c341bbdde075f97f7947e9160 Mon Sep 17 00:00:00 2001
From: Henry Dong <donghenry@google.com>
Date: Fri, 27 Mar 2026 05:41:28 +0000
Subject: [PATCH 53/63] ANDROID: vendor_hooks: Add hook for sync mmap readahead

Add vendor hooks to precisely control sync readahead based on
various factors, such as filesystem type, file type, process
priority, and other considerations.

Bug: 496697696
Bug: 498398295

Change-Id: I5842f1a313a80bebf87fb810e31d6d2d44b96cab
Signed-off-by: Henry Dong <donghenry@google.com>
(cherry picked from commit 5dd2a858247711696bafb2172b693aee6d0b3c31)
---
 drivers/android/vendor_hooks.c | 1 +
 include/trace/hooks/mm.h       | 3 +++
 mm/filemap.c                   | 5 +++++
 3 files changed, 9 insertions(+)

diff --git a/drivers/android/vendor_hooks.c b/drivers/android/vendor_hooks.c
index a06eda6682b1..9fb80c4fe068 100644
--- a/drivers/android/vendor_hooks.c
+++ b/drivers/android/vendor_hooks.c
@@ -435,6 +435,7 @@ EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_rcu_boost_end);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_lruvec_add_folio);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_lruvec_del_folio);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_do_async_mmap_readahead);
+EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_do_sync_mmap_readahead);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_mm_free_page);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_page_cache_ra_unbounded);
 EXPORT_TRACEPOINT_SYMBOL_GPL(android_vh_ra_alloc_retry);
diff --git a/include/trace/hooks/mm.h b/include/trace/hooks/mm.h
index 7e5a8410b720..6e3859306b85 100644
--- a/include/trace/hooks/mm.h
+++ b/include/trace/hooks/mm.h
@@ -357,6 +357,9 @@ DECLARE_HOOK(android_vh_add_lazyfree_bypass,
 DECLARE_HOOK(android_vh_do_async_mmap_readahead,
 	TP_PROTO(struct vm_fault *vmf, struct folio *folio, bool *skip),
 	TP_ARGS(vmf, folio, skip));
+DECLARE_HOOK(android_vh_do_sync_mmap_readahead,
+	TP_PROTO(struct vm_fault *vmf, bool *skip),
+	TP_ARGS(vmf, skip));
 DECLARE_HOOK(android_vh_mm_free_page,
 	TP_PROTO(struct page *page),
 	TP_ARGS(page));
diff --git a/mm/filemap.c b/mm/filemap.c
index 26e7d5794d6a..70ff22a1ffad 100644
--- a/mm/filemap.c
+++ b/mm/filemap.c
@@ -3291,6 +3291,11 @@ static struct file *do_sync_mmap_readahead(struct vm_fault *vmf)
 	struct file *fpin = NULL;
 	unsigned long vm_flags = vmf->vma->vm_flags;
 	unsigned short mmap_miss;
+	bool skip = false;
+
+	trace_android_vh_do_sync_mmap_readahead(vmf, &skip);
+	if (skip)
+		return fpin;
 
 #ifdef CONFIG_TRANSPARENT_HUGEPAGE
 	/* Use the readahead code, even if readahead is disabled */
-- 
2.50.1

